### PR TITLE
[Merged by Bors] - Kiln mev boost

### PR DESCRIPTION
**Pull request by realbigsean**
## Issue Addressed

MEV boost compatibility

## Proposed Changes

See #2987

## Additional Info

This is blocked on the stabilization of a couple specs, [here](https://github.com/ethereum/beacon-APIs/pull/194) and [here](https://github.com/flashbots/mev-boost/pull/20).

Additional TODO's and outstanding questions

- [ ] MEV boost JWT Auth
- [ ] Will `builder_proposeBlindedBlock` return the revealed payload for the BN to propogate
- [ ] Should we remove `private-tx-proposals` flag and communicate BN <> VC with blinded blocks by default once these endpoints enter the beacon-API's repo? This simplifies merge transition logic.

### CODE REVIEW

File: beacon_node/execution_layer/src/engine_api/json_structures.rs
Comment (realbigsean) [inline beacon_node/execution_layer/src/engine_api/json_structures.rs:403]: This is no longer necessary since we removed the union from the transaction type in the consensus spec

Comment (michaelsproul) [reply beacon_node/execution_layer/src/engine_api/json_structures.rs:403]: Nice

File: beacon_node/execution_layer/src/lib.rs
Comment (michaelsproul) [inline beacon_node/execution_layer/src/lib.rs]: I think this log message should be lifted out so that it occurs in both `Blinded`/`Full` cases

File: beacon_node/http_api/src/lib.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs:1065]: Just making a note here that we only support merge blocks on this endpoint currently, but the spec as currently written seems to imply we should support pre-Bellatrix blocks too

File: beacon_node/execution_layer/src/engine_api/http.rs
Comment (michaelsproul) [inline beacon_node/execution_layer/src/engine_api/http.rs]: Can we not use the `SignedBeaconBlock<T, BlindedTransactions>` type directly here, and do away with the separate `BlindedBeaconBlock` types? Is there something about (de)serialization that breaks?

File: validator_client/src/validator_store.rs
Comment (michaelsproul) [inline validator_client/src/validator_store.rs]: we may as well import this through `types::Transactions`

File: validator_client/src/block_service.rs
Comment (michaelsproul) [inline validator_client/src/block_service.rs]: I think we can use `slot` from further back in the function and avoid the `unwrap`

File: validator_client/src/block_service.rs
Comment (michaelsproul) [inline validator_client/src/block_service.rs]: this seems like a good way to handle the transition, although I would have thought the check should be `now >= merge_slot`?

what problems do you foresee?

**Review by michaelsproul — CHANGES_REQUESTED** [review]
This is looking really neat!

I'd like to get this merged ASAP (post v2.1.4) so that we can start re-structuring the database around blinded blocks (i.e. store blinded blocks in the DB, with separate execution payloads that get erased on finalization).

I think so long as we have the types nailed down and we maintain backwards compatibility with the existing APIs when `--private-tx-proposals` is turned off then I think we can make further tweaks to the blinded block proposal flow after merging.

> Should we remove private-tx-proposals flag and communicate BN <> VC with blinded blocks by default once these endpoints enter the beacon-API's repo? This simplifies merge transition logic.

I think we should keep the opt-in flag so that the VC retains the ability to sign full blocks, and so we can maintain backwards compatibility with BNs that lack the `blinded_blocks` APIs leading up to the merge.

----

As far as I can tell the main changes that might occur can be handled with small tweaks to this PR after merging:

* Add support for pre-Bellatrix blocks to the `/blinded_blocks` HTTP endpoints
* Re-structure to support mev-boost as a "side car" rather than middleware, as per https://github.com/flashbots/mev-boost/pull/20/files#r788082526. I'm not sure what the appetite is for this change and whether it'll go through

File: beacon_node/execution_layer/src/engine_api/http.rs
Comment (realbigsean) [reply beacon_node/execution_layer/src/engine_api/http.rs]: The issue I ran into was the name of the field `transactions` (signed block's execution payload) vs `transactions_root` (signed blinded block's execution payload header). We could maybe figure out a way to do a conditional rename but I don't think there's a way to do it within `serde_derive`.  You can [alias](https://serde.rs/field-attrs.html#alias) on deserialization, but there's nothing like a `rename_if` for serialization.

This is actually making me realize we'd be slightly off-spec as-is because our BN<>VC communication would have the field name `transactions` if we don't do this same transformation in that API.

File: validator_client/src/block_service.rs
Comment (realbigsean) [reply validator_client/src/block_service.rs]: I was generally worried about outsourcing payload construction at transition, which paul addressed [here](https://github.com/sigp/lighthouse/pull/2987#issuecomment-1028487109) (I haven't implented his suggestion) 

But if this endpoint will end up serving pre-Bellatrix blocks, it might be better to remove any slot logic here and just add logic for "only outsource payload construction if we have finalized" to the beacon node

File: beacon_node/execution_layer/src/lib.rs
Comment (realbigsean) [reply beacon_node/execution_layer/src/lib.rs]: Whoops there's a couple mistakes here.. I accidentally deleted the full block log here, probably during a merge with unstable. And the two logs are slightly different because the endpoints they hit have different names. This one should be `builder_getPayloadHeader` and the other one should be `engine_getPayload`

**Comment by realbigsean** [comment]
Raised an issue [here](https://github.com/sigp/lighthouse/issues/3091) to track additional changes outside of this PR to try to get the new types merged more quickly

File: beacon_node/http_api/src/lib.rs
Comment (michaelsproul) [reply beacon_node/http_api/src/lib.rs:1065]: This is tracked by #3091

File: beacon_node/execution_layer/src/engine_api/http.rs
Comment (michaelsproul) [reply beacon_node/execution_layer/src/engine_api/http.rs]: I've opened a PR to this branch that uses `serde(flatten)` to unify the types, let me know if you think it's flexible enough: https://github.com/realbigsean/lighthouse/pull/20

File: validator_client/src/block_service.rs
Comment (michaelsproul) [reply validator_client/src/block_service.rs]: Ah yeah, that sounds sensible. I'm happy to come back to that after merging an MVP version of this PR.

**Comment by realbigsean** [comment]
@michaelsproul so I've tested this now and gotten it working using this repo: https://github.com/realbigsean/lighthouse-kiln-mev-boost

An outstanding limitation is that there is no relay running that can handle kiln blocks so I'm working on a local mock relay in a different branch. 

I had to make some updates to get this working here. The main thing is that mev-boost no longer handles any of our `eth_syncing` calls, so we can no longer incorporate it into our usual fallback routine, where we repeatedly make those calls and update a local picture of it's sync state.  I am going to make some larger changes in a separate PR, to separate the builder API from the engine API and untangle the fallback routine of the two. In this PR is a minimal, workable, not-ideal solution where I've just removed any `eth_syncing` calls but in doing so, removed builder to builder fallback logic. 

Additionally there's a couple of ssz static tests failing.. I'm thinking it might be because we need something like `#[ssz(flatten)]` for the new payload wrapper?

**Comment by realbigsean** [comment]
Oh you know it's actually failing when it tries to read from the yaml file. I think `#[serde(flatten)]` and `#[serde(transparent)]` are both removing a layer of the struct hierarchy so we end up with `BeaconBlockBody` including `ExecutionPayload` fields directly

**Comment by realbigsean** [comment]
Ok I think this should fix it https://github.com/sigp/lighthouse/pull/3062/commits/c36bafca4333edfa4c16a8953127264d22892999

🤞

**Comment by michaelsproul** [comment]
Oh yeah, good catch. Looks like that fixed it!

**Comment by realbigsean** [comment]
I made some more parts of the `execution_layer` crate public for use in the mock relay I'm working on [here](https://github.com/realbigsean/mock-relay)

File: beacon_node/http_api/src/lib.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs:1096]: This `unwrap` isn't ideal, but I note that you've fixed it in #3134 and it isn't reachable from non-blinded block proposal routines, so I think we're good :+1:

**Review by michaelsproul — APPROVED** [review]
I'm happy to merge this!

**Comment by michaelsproul** [comment]
Merging this so PRs built on top can target unstable

bors r+

File: beacon_node/http_api/src/lib.rs
Comment (realbigsean) [reply beacon_node/http_api/src/lib.rs:1096]: Whoops, oversight on my part!

**Pull request closed without merging**
